File: frontend/services/shifts.ts
import { apiFetch } from "@/services/api/client";
import type {
  Shift,
  ShiftTemplate,
  OpenShiftClaim,
  ShiftSwapRequest,
  LeaveRequest,
  StaffAvailability,
  AttendanceRecord,
  AttendanceRules,
  TimesheetSummaryRow,
  PaginatedResponse,
} from "@/types";

// ── Shift Templates ────────────────────────────────────────────────────────────

export function listShiftTemplates(params?: { location_id?: string }): Promise<ShiftTemplate[]> {
  const q = new URLSearchParams();
  if (params?.location_id) q.set("location_id", params.location_id);
  const qs = q.toString();
  return apiFetch(`/api/v1/shifts/templates${qs ? `?${qs}` : ""}`);
}

export function createShiftTemplate(body: {
  name: string;
  role?: string | null;
  start_time: string;
  end_time: string;
  days_of_week: number[];
  location_id?: string | null;  // null = org-wide (admin only)
  is_active?: boolean;
}): Promise<ShiftTemplate> {
  return apiFetch("/api/v1/shifts/templates", { method: "POST", body: JSON.stringify(body) });
}

export function updateShiftTemplate(
  id: string,
  body: Partial<{
    name: string;
    role: string | null;
    start_time: string;
    end_time: string;
    days_of_week: number[];
    location_id: string;
    is_active: boolean;
  }>
): Promise<ShiftTemplate> {
  return apiFetch(`/api/v1/shifts/templates/${id}`, { method: "PUT", body: JSON.stringify(body) });
}

export function deleteShiftTemplate(id: string): Promise<{ ok: boolean }> {
  return apiFetch(`/api/v1/shifts/templates/${id}`, { method: "DELETE" });
}

export function bulkGenerateShifts(
  templateId: string,
  body: { date_from: string; date_to: string; location_id?: string }
): Promise<{ shifts_created: number }> {
  return apiFetch(`/api/v1/shifts/templates/${templateId}/generate`, {
    method: "POST",
    body: JSON.stringify({ template_id: templateId, ...body }),
  });
}

// ── Shifts ─────────────────────────────────────────────────────────────────────

export function listShifts(params?: {
  location_id?: string;
  user_id?: string;
  status?: string;
  from_date?: string;
  to_date?: string;
  page?: number;
  page_size?: number;
}): Promise<PaginatedResponse<Shift>> {
  const q = new URLSearchParams();
  if (params?.location_id) q.set("location_id", params.location_id);
  if (params?.user_id)     q.set("user_id", params.user_id);
  if (params?.status)      q.set("status", params.status);
  if (params?.from_date)   q.set("from_date", params.from_date);
  if (params?.to_date)     q.set("to_date", params.to_date);
  q.set("page", String(params?.page ?? 1));
  q.set("page_size", String(params?.page_size ?? 50));
  return apiFetch(`/api/v1/shifts/?${q}`);
}

export function getShift(id: string): Promise<Shift> {
  return apiFetch(`/api/v1/shifts/${id}`);
}

export function createShift(body: {
  location_id: string;
  role?: string | null;
  start_at: string;
  end_at: string;
  assigned_to_user_id?: string | null;
  template_id?: string | null;
  notes?: string | null;
  is_open_shift?: boolean;
  status?: string;
}): Promise<Shift> {
  return apiFetch("/api/v1/shifts/", { method: "POST", body: JSON.stringify(body) });
}

export function updateShift(
  id: string,
  body: Partial<{
    location_id: string;
    role: string | null;
    start_at: string;
    end_at: string;
    assigned_to_user_id: string | null;
    notes: string | null;
    is_open_shift: boolean;
    status: string;
    cancellation_reason: string | null;
  }>
): Promise<Shift> {
  return apiFetch(`/api/v1/shifts/${id}`, { method: "PUT", body: JSON.stringify(body) });
}

export function deleteShift(id: string): Promise<{ ok: boolean }> {
  return apiFetch(`/api/v1/shifts/${id}`, { method: "DELETE" });
}

export function publishShifts(shiftIds: string[]): Promise<{ published: number }> {
  return apiFetch("/api/v1/shifts/publish", {
    method: "POST",
    body: JSON.stringify({ shift_ids: shiftIds }),
  });
}

export function publishBulk(body: {
  filter_type: "location" | "role" | "individual";
  location_id?: string | null;
  role?: string | null;
  user_id?: string | null;
  week_start?: string;
  week_end?: string;
}): Promise<{ published: number }> {
  return apiFetch("/api/v1/shifts/publish/bulk", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

// ── Open Shift Claims ─────────────────────────────────────────────────────────

export function claimShift(shiftId: string): Promise<OpenShiftClaim> {
  return apiFetch(`/api/v1/shifts/${shiftId}/claim`, { method: "POST" });
}

export function listClaims(params?: {
  shift_id?: string;
  status?: string;
}): Promise<OpenShiftClaim[]> {
  const q = new URLSearchParams();
  if (params?.shift_id) q.set("shift_id", params.shift_id);
  if (params?.status)   q.set("status", params.status);
  const qs = q.toString();
  return apiFetch(`/api/v1/shifts/claims${qs ? `?${qs}` : ""}`);
}

export function respondToClaim(
  claimId: string,
  body: { action: "approve" | "reject"; manager_note?: string }
): Promise<OpenShiftClaim> {
  return apiFetch(`/api/v1/shifts/claims/${claimId}/respond`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

// ── Swap Requests ─────────────────────────────────────────────────────────────

export function listSwapRequests(params?: { status?: string }): Promise<ShiftSwapRequest[]> {
  const q = new URLSearchParams();
  if (params?.status) q.set("status", params.status);
  const qs = q.toString();
  return apiFetch(`/api/v1/shifts/swaps${qs ? `?${qs}` : ""}`);
}

export function createSwapRequest(body: {
  shift_id: string;
  target_user_id?: string | null;
  target_shift_id?: string | null;
}): Promise<ShiftSwapRequest> {
  return apiFetch("/api/v1/shifts/swaps", { method: "POST", body: JSON.stringify(body) });
}

export function respondToSwap(
  swapId: string,
  body: { action: string; rejection_reason?: string }
): Promise<ShiftSwapRequest> {
  return apiFetch(`/api/v1/shifts/swaps/${swapId}/respond`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

// ── Leave Requests ────────────────────────────────────────────────────────────

export function listLeaveRequests(params?: {
  status?: string;
  user_id?: string;
  page?: number;
  page_size?: number;
}): Promise<PaginatedResponse<LeaveRequest>> {
  const q = new URLSearchParams();
  if (params?.status)  q.set("status", params.status);
  if (params?.user_id) q.set("user_id", params.user_id);
  q.set("page", String(params?.page ?? 1));
  q.set("page_size", String(params?.page_size ?? 20));
  return apiFetch(`/api/v1/shifts/leave?${q}`);
}

export function createLeaveRequest(body: {
  leave_type: string;
  start_date: string;
  end_date: string;
  reason?: string | null;
}): Promise<LeaveRequest> {
  return apiFetch("/api/v1/shifts/leave", { method: "POST", body: JSON.stringify(body) });
}

export function respondToLeave(
  leaveId: string,
  body: { action: "approve" | "reject" }
): Promise<LeaveRequest> {
  return apiFetch(`/api/v1/shifts/leave/${leaveId}/respond`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

// ── Availability ──────────────────────────────────────────────────────────────

export function getMyAvailability(params?: { user_id?: string }): Promise<StaffAvailability[]> {
  const q = new URLSearchParams();
  if (params?.user_id) q.set("user_id", params.user_id);
  const qs = q.toString();
  return apiFetch(`/api/v1/shifts/availability${qs ? `?${qs}` : ""}`);
}

export function setAvailability(body: {
  day_of_week: number;
  available_from: string;
  available_to: string;
  is_available?: boolean;
  effective_from?: string | null;
  effective_to?: string | null;
}): Promise<StaffAvailability> {
  return apiFetch("/api/v1/shifts/availability", { method: "POST", body: JSON.stringify(body) });
}

// ── Attendance ────────────────────────────────────────────────────────────────

export function clockIn(body: {
  shift_id?: string | null;
  location_id: string;
  clock_in_method?: string;
  latitude?: number;
  longitude?: number;
}): Promise<AttendanceRecord> {
  return apiFetch("/api/v1/shifts/attendance/clock-in", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function clockOut(body: {
  attendance_id: string;
  latitude?: number;
  longitude?: number;
}): Promise<AttendanceRecord> {
  return apiFetch("/api/v1/shifts/attendance/clock-out", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function listAttendance(params?: {
  user_id?: string;
  location_id?: string;
  from_date?: string;
  to_date?: string;
  status?: string;
  page?: number;
  page_size?: number;
}): Promise<PaginatedResponse<AttendanceRecord>> {
  const q = new URLSearchParams();
  if (params?.user_id)     q.set("user_id", params.user_id);
  if (params?.location_id) q.set("location_id", params.location_id);
  if (params?.from_date)   q.set("from_date", params.from_date);
  if (params?.to_date)     q.set("to_date", params.to_date);
  if (params?.status)      q.set("status", params.status);
  q.set("page", String(params?.page ?? 1));
  q.set("page_size", String(params?.page_size ?? 50));
  return apiFetch(`/api/v1/shifts/attendance?${q}`);
}

export function getMyAttendance(params?: {
  from_date?: string;
  to_date?: string;
}): Promise<AttendanceRecord[]> {
  const q = new URLSearchParams();
  if (params?.from_date) q.set("from_date", params.from_date);
  if (params?.to_date)   q.set("to_date", params.to_date);
  q.set("page_size", "200");
  return (apiFetch(`/api/v1/shifts/attendance?${q}`) as Promise<PaginatedResponse<AttendanceRecord>>).then(
    (r) => r.items
  );
}

export function managerOverride(body: {
  user_id: string;
  shift_id?: string | null;
  location_id: string;
  clock_in_at: string;
  clock_out_at?: string | null;
  note: string;
}): Promise<AttendanceRecord> {
  return apiFetch("/api/v1/shifts/attendance/override", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function getTimesheetSummary(params?: {
  week_start?: string;
  location_id?: string;
  user_id?: string;
}): Promise<TimesheetSummaryRow[]> {
  const q = new URLSearchParams();
  if (params?.week_start)  q.set("week_start", params.week_start);
  if (params?.location_id) q.set("location_id", params.location_id);
  if (params?.user_id)     q.set("user_id", params.user_id);
  const qs = q.toString();
  return apiFetch(`/api/v1/shifts/attendance/timesheet${qs ? `?${qs}` : ""}`);
}

export function getMyTimesheet(params?: { week_start?: string }): Promise<{
  records: AttendanceRecord[];
  summary: { total_hours: number; break_hours: number; worked_hours: number; regular_hours: number; overtime_hours: number; late_count: number };
}> {
  const q = new URLSearchParams();
  if (params?.week_start) q.set("week_start", params.week_start);
  const qs = q.toString();
  return apiFetch(`/api/v1/shifts/attendance/my-timesheet${qs ? `?${qs}` : ""}`);
}

// ── Attendance Rules ──────────────────────────────────────────────────────────

export function getAttendanceRules(): Promise<AttendanceRules> {
  return apiFetch("/api/v1/shifts/rules");
}

export function updateAttendanceRules(body: Partial<{
  late_threshold_mins: number;
  early_departure_threshold_mins: number;
  overtime_threshold_hours: number;
  weekly_overtime_threshold_hours: number;
  break_duration_mins: number;
}>): Promise<AttendanceRules> {
  return apiFetch("/api/v1/shifts/rules", { method: "PUT", body: JSON.stringify(body) });
}

// ── Breaks ────────────────────────────────────────────────────────────────────

export function startBreak(body: {
  attendance_id: string;
  break_type?: "meal" | "rest" | "other";
}): Promise<{ id: string; break_start_at: string; break_type: string }> {
  return apiFetch("/api/v1/shifts/attendance/break/start", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function endBreak(body: {
  attendance_id: string;
}): Promise<{ id: string; break_end_at: string; duration_minutes: number }> {
  return apiFetch("/api/v1/shifts/attendance/break/end", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function getBreakStatus(attendanceId: string): Promise<{
  on_break: boolean;
  active_break: { id: string; break_start_at: string; break_type: string } | null;
  breaks: Array<{ id: string; break_start_at: string; break_end_at: string | null; duration_minutes: number | null; break_type: string }>;
  total_break_minutes: number;
}> {
  return apiFetch(`/api/v1/shifts/attendance/break/status?attendance_id=${attendanceId}`);
}

// ── AI Schedule ───────────────────────────────────────────────────────────────

export function generateAISchedule(body: {
  location_id: string;
  week_start: string;
  notes?: string;
}): Promise<{ shifts_created: number; warnings: string[] }> {
  return apiFetch("/api/v1/shifts/ai/generate-schedule", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

File: backend/models/shifts.py
from __future__ import annotations

from datetime import date, time
from typing import Optional
from pydantic import BaseModel


# ── Shift Templates ────────────────────────────────────────────────────────────

class CreateShiftTemplateRequest(BaseModel):
    name: str
    role: Optional[str] = None
    start_time: str          # "HH:MM"
    end_time: str            # "HH:MM"
    days_of_week: list[int]  # 0=Mon … 6=Sun
    location_id: Optional[str] = None   # None = org-wide (admin only)
    is_active: bool = True


class UpdateShiftTemplateRequest(BaseModel):
    name: Optional[str] = None
    role: Optional[str] = None
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    days_of_week: Optional[list[int]] = None
    location_id: Optional[str] = None
    is_active: Optional[bool] = None


# ── Shifts ─────────────────────────────────────────────────────────────────────

class CreateShiftRequest(BaseModel):
    location_id: str
    role: Optional[str] = None
    start_at: str                       # ISO datetime string
    end_at: str                         # ISO datetime string
    assigned_to_user_id: Optional[str] = None
    template_id: Optional[str] = None
    notes: Optional[str] = None
    is_open_shift: bool = False
    status: str = "draft"


class UpdateShiftRequest(BaseModel):
    location_id: Optional[str] = None
    role: Optional[str] = None
    start_at: Optional[str] = None
    end_at: Optional[str] = None
    assigned_to_user_id: Optional[str] = None
    notes: Optional[str] = None
    is_open_shift: Optional[bool] = None
    status: Optional[str] = None
    cancellation_reason: Optional[str] = None


class BulkGenerateShiftsRequest(BaseModel):
    template_id: str
    date_from: date
    date_to: date
    location_id: Optional[str] = None  # required when template is org-wide (no location_id)


class ShiftAssignment(BaseModel):
    shift_id: str
    user_id: Optional[str] = None
    is_open_shift: bool = False


class BulkAssignRequest(BaseModel):
    assignments: list[ShiftAssignment]


class PublishShiftsRequest(BaseModel):
    shift_ids: list[str]


class BulkPublishRequest(BaseModel):
    filter_type: str                    # "location" | "role" | "individual"
    location_id: Optional[str] = None  # for filter_type="location"
    role: Optional[str] = None         # for filter_type="role"
    user_id: Optional[str] = None      # for filter_type="individual"
    week_start: Optional[str] = None   # ISO date "YYYY-MM-DD"
    week_end: Optional[str] = None     # ISO date "YYYY-MM-DD"


# ── Open Shifts ────────────────────────────────────────────────────────────────

class ClaimShiftRequest(BaseModel):
    shift_id: str


class RespondToClaimRequest(BaseModel):
    action: str   # 'approve' | 'reject'
    manager_note: Optional[str] = None


# ── Swap Requests ──────────────────────────────────────────────────────────────

class CreateSwapRequest(BaseModel):
    shift_id: str
    target_user_id: Optional[str] = None
    target_shift_id: Optional[str] = None


class RespondToSwapRequest(BaseModel):
    action: str                         # 'accept' | 'decline' | 'approve' | 'reject'
    rejection_reason: Optional[str] = None


# ── Leave ──────────────────────────────────────────────────────────────────────

class CreateLeaveRequest(BaseModel):
    leave_type: str   # 'annual' | 'sick' | 'emergency' | 'unpaid' | 'other'
    start_date: date
    end_date: date
    reason: Optional[str] = None


class RespondToLeaveRequest(BaseModel):
    action: str   # 'approve' | 'reject'


# ── Availability ───────────────────────────────────────────────────────────────

class SetAvailabilityRequest(BaseModel):
    day_of_week: int        # 0=Mon … 6=Sun
    available_from: str     # "HH:MM"
    available_to: str       # "HH:MM"
    is_available: bool = True
    effective_from: Optional[str] = None   # date string "YYYY-MM-DD"
    effective_to: Optional[str] = None     # date string "YYYY-MM-DD"


# ── Attendance ─────────────────────────────────────────────────────────────────

class ClockInRequest(BaseModel):
    shift_id: Optional[str] = None
    location_id: str
    clock_in_method: str = "gps"
    latitude: Optional[float] = None
    longitude: Optional[float] = None


class ClockOutRequest(BaseModel):
    attendance_id: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None


class ManagerOverrideRequest(BaseModel):
    user_id: str
    shift_id: Optional[str] = None
    location_id: str
    clock_in_at: str   # ISO datetime string
    clock_out_at: Optional[str] = None
    note: str


# ── Attendance Rules ───────────────────────────────────────────────────────────

class UpdateAttendanceRulesRequest(BaseModel):
    late_threshold_mins: Optional[int] = None
    early_departure_threshold_mins: Optional[int] = None
    overtime_threshold_hours: Optional[float] = None
    weekly_overtime_threshold_hours: Optional[float] = None
    break_duration_mins: Optional[int] = None


# ── Breaks ─────────────────────────────────────────────────────────────────────

class StartBreakRequest(BaseModel):
    attendance_id: str
    break_type: str = "rest"   # "meal" | "rest" | "other"


class EndBreakRequest(BaseModel):
    attendance_id: str


# ── AI Schedule ────────────────────────────────────────────────────────────────

class GenerateScheduleRequest(BaseModel):
    location_id: str
    week_start: date
    notes: Optional[str] = None
